MaterialProperties Refactoring step 2

yujiex opened the pull request:
Pull request overview
---------------------
This is stage two of the refactoring of Materials: splitting the `MaterialProperties` class into a `MaterialBase` class containing Group and Name and a `MaterialChild` class which includes everything else. This facilitates the further split into `MaterialChild` class into more specific subclasses.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

yujiex committed:
min(max(xx, xx), xx) change to std::clamp

yujiex committed:
Step 2 split MaterialProperties into MaterialBase, MaterialChild

yujiex committed:
Merge remote-tracking branch 'origin/develop' into dynamicCoatingWallRoof

yujiex committed:
revert clamp in SingleDuct

yujiex committed:
clang-format

amirroth commented on src/EnergyPlus/Construction.cc:
```diff
@@ -215,47 +215,48 @@ void ConstructionProps::calculateTransferFunction(EnergyPlusData &state, bool &E
         // Obtain thermal properties from the Material derived type
 
         auto *thisMaterial = state.dataMaterial->Material(CurrentLayer);
+        auto *thisMaterialChild = dynamic_cast<Material::MaterialChild *>(thisMaterial);
```
You can keep calling this `thisMaterial`. That will minimize the changes and be shorter also.

amirroth commented on src/EnergyPlus/DataHeatBalance.cc:
```diff
@@ -401,29 +401,36 @@ void CheckAndSetConstructionProperties(EnergyPlusData &state,
     for (Layer = 1; Layer <= TotLayers; ++Layer) {
         MaterNum = thisConstruct.LayerPoint(Layer);
         auto const *thisMaterial = state.dataMaterial->Material(MaterNum);
+        auto const *thisMaterialChild = dynamic_cast<const Material::MaterialChild *>(thisMaterial);
```
Not need to change the name of the shortcut reference.

amirroth commented on src/EnergyPlus/EcoRoofManager.cc:
```diff
@@ -705,16 +707,17 @@ namespace EcoRoofManager {
         using namespace DataEnvironment;
 
         auto const *thisMat = state.dataMaterial->Material(state.dataConstruction->Construct(ConstrNum).LayerPoint(1));
+        auto const *thisMatChild = dynamic_cast<const Material::MaterialChild *>(thisMat);
```
Again, just `auto const *thisMat = dynamic_cast<const Material::MaterialChild *>(state.dataMaterial->Material(state.dataConstruction->Construct(ConstrNum).LayerPoint(1));`  is cleaner and easier to review. If you want to shorten the line, then `auto &thisConstruction` is probably a better idea.

yujiex replied:
will change to this format instead

amirroth commented on src/EnergyPlus/HeatBalFiniteDiffManager.cc:
```diff
@@ -236,6 +236,7 @@ namespace HeatBalFiniteDiffManager {
                 // Load the material derived type from the input data.
                 MaterNum = UtilityRoutines::FindItemInPtrList(MaterialNames(1), state.dataMaterial->Material);
                 auto const *thisMaterial = state.dataMaterial->Material(MaterNum);
+                auto const *thisMaterialChild = dynamic_cast<const Material::MaterialChild *>(thisMaterial);
```
`dynamic_cast` will crash I believe if `thisMaterial` is `nullptr`. This shortcut has to come after the `if (MaterNum == 0)` test.

yujiex replied:
indeed. just moved it to after the MaterialNum == 0 check

amirroth commented on src/EnergyPlus/HeatBalanceManager.cc:
```diff
@@ -1758,11 +1758,12 @@ namespace HeatBalanceManager {
-                    if (state.dataMaterial->Material(thisConstruct.LayerPoint(Layer))->Group == Material::MaterialGroup::WindowGlass) ++iMatGlass;
-                    MaterialLayerGroup = state.dataMaterial->Material(thisConstruct.LayerPoint(Layer))->Group;
+                    auto const *thisMaterial = state.dataMaterial->Material(thisConstruct.LayerPoint(Layer));
+                    auto const *thisMaterialChild = dynamic_cast<const Material::MaterialChild *>(thisMaterial);
```
If you only need the `Group` field, you don't have to do the `dynamic cast`. That should accessible from `MaterialBase`.

amirroth commented on src/EnergyPlus/HeatBalanceManager.cc:
```diff
@@ -4550,18 +4564,21 @@ namespace HeatBalanceManager {
             for (IGlSys = 1; IGlSys <= NGlSys; ++IGlSys) {
                 for (IGap = 1; IGap <= NGaps(IGlSys); ++IGap) {
                     ++MaterNum;
+                    delete state.dataMaterial->Material(MaterNum);
```
Don't do this `delete`. Just don't create `MaterialBase` objects to begin with.

yujiex committed:
move material shortcut to after MaterNum == 0 check

yujiex committed:
Fix mat2Child shortcut cast from mat

yujiex committed:
change function args and member init to matChild

yujiex committed:
name change: thisMaterialChild to thisMaterial

yujiex committed:
add nullptr assertion

yujiex committed:
Fix Window5Data reading issue (causing integration err)

yujiex committed:
clang-format

yujiex committed:
clang-format in unit test

yujiex committed:
Remove redundant upcast and downcast

yujiex committed:
Merge remote-tracking branch 'origin/develop' into dynamicCoatingWallRoof

yujiex committed:
clang-format

yujiex committed:
more cleanup

yujiex committed:
Revert "min(max(xx, xx), xx) change to std::clamp"

This reverts commit b17cf5b99d4a3d95bdc128e85a083118a52da7e9.

Myoldmopar commented:
@yujiex so this PR is just another level of refactoring, right?  If so can you change the PR name to make it more accurate with the refactoring going on.  Are you done working here or is there additional work to do?

yujiex commented:
> @yujiex so this PR is just another level of refactoring, right? If so can you change the PR name to make it more accurate with the refactoring going on. Are you done working here or is there additional work to do?

Hi @Myoldmopar, I just changed the PR name. It is the second step of the MaterialProperties refactoring. I finished the changes I'd like to make here but there are a couple of regression diffs I haven't been able to resolve yet. So I switched to work on the coating feature. I was planning to go back to this branch after the coating feature is done.

yujiex committed:
Merge remote-tracking branch 'origin/develop' into dynamicCoatingWallRoof

Myoldmopar commented:
@yujiex now that IO freeze is over, please advise your plan on this branch.  If you intend to push on it over the next couple weeks, we'll keep it marked for the 23.1 milestone.  If not, we'll push to the next release so it doesn't get in the way during reviews.

yujiex commented:
> @yujiex now that IO freeze is over, please advise your plan on this branch. If you intend to push on it over the next couple weeks, we'll keep it marked for the 23.1 milestone. If not, we'll push to the next release so it doesn't get in the way during reviews.

Hi @Myoldmopar, I am working on this branch right now. I had some regression diffs to fix. Currently I'm rewinding some steps to see where caused the difference. I think I'll be able to finish this within the next couple of weeks before the 23.1 release.

yujiex committed:
Cherrypick the good commit without reg diff f3e6aefe6a

amirroth commented on src/EnergyPlus/MicroturbineElectricGenerator.cc:
```diff
@@ -1354,7 +1354,7 @@ void MTGeneratorSpecs::CalcMTGeneratorModel(EnergyPlusData &state,
 
         //     Calculate operating power output (gross)
-        elecPowerGenerated = min(max(0.0, MyLoad + ancillaryPowerRate), FullLoadPowerOutput);
+        elecPowerGenerated = std::clamp(0.0, MyLoad + ancillaryPowerRate, FullLoadPowerOutput);
```
The value you are trying to clamp goes first.

amirroth commented on src/EnergyPlus/Material.hh:
```diff
@@ -132,17 +132,29 @@ namespace Material {
-        std::string Name;              // Name of material layer
-        Material::MaterialGroup Group; // Material group type (see Material Parameters above.  Currently
+        std::string Name = "";                                            // Name of material layer
+        Material::MaterialGroup Group = Material::MaterialGroup::Invalid; // Material group type (see Material Parameters above.  Currently
```
I wonder if it makes sense to move some other commonly used members into MaterialBase to reduce the frequency with which you have to downcast to a child class. Things like `ROnly`, `Resistance`, maybe other things too. What do you think?

yujiex replied:
Yeah, maybe we can make the material base a bit larger adding a few more frequently-used members. I was still debugging the regression diffs. I was going back and rewinding some of the order changes in the splits: I renamed the MaterialProperties into MaterialBase and created an empty MaterialChild class inheriting from MaterialBase. Then gradually moving members from MaterialBase to MaterialChild. So far seems good. Still in the process of working my way through the members.

amirroth commented on src/EnergyPlus/Material.cc:
```diff
@@ -219,7 +219,7 @@ void GetMaterialData(EnergyPlusData &state, bool &ErrorsFound) // set to true if
 
     for (int i = 1; i <= state.dataMaterial->TotMaterials; i++) {
-        MaterialProperties *p = new MaterialProperties;
+        MaterialBase *p = new MaterialBase;
```
Don't need to do this.

yujiex replied:
Will remove this

amirroth commented on src/EnergyPlus/Material.cc:
```diff
@@ -257,7 +257,10 @@ void GetMaterialData(EnergyPlusData &state, bool &ErrorsFound) // set to true if
 
             // Load the material derived type from the input data.
-            auto thisMaterial = state.dataMaterial->Material(MaterNum);
+            delete state.dataMaterial->Material(MaterNum);
```
Don't need to do this.

yujiex replied:
right, will not need to delete when the previous new MaterialBase is removed

amirroth commented on src/EnergyPlus/Material.cc:
```diff
@@ -257,7 +257,10 @@ void GetMaterialData(EnergyPlusData &state, bool &ErrorsFound) // set to true if
-            auto thisMaterial = state.dataMaterial->Material(MaterNum);
+            delete state.dataMaterial->Material(MaterNum);
+            state.dataMaterial->Material(MaterNum) = new Material::MaterialChild;
+            auto *thisMaterial = dynamic_cast<Material::MaterialChild *>(state.dataMaterial->Material(MaterNum));
```
Don't need to downcast if you assign `thisMaterial` to `new Material::MaterialChild first`.

yujiex replied:
right right. That's better

amirroth commented on src/EnergyPlus/Material.cc:
```diff
@@ -289,7 +292,10 @@ void GetMaterialData(EnergyPlusData &state, bool &ErrorsFound) // set to true if
     if (TotFfactorConstructs + TotCfactorConstructs >= 1) {
         ++MaterNum;
-        auto *thisMaterial = state.dataMaterial->Material(MaterNum);
+        delete state.dataMaterial->Material(MaterNum);
```
Same set of comments.

yujiex replied:
got it

yujiex committed:
Merge remote-tracking branch 'origin/develop' into dynamicCoatingWallRoof

yujiex committed:
Revert "Step 2 split MaterialProperties into MaterialBase, MaterialChild"

This reverts commit 706fc8707520f9b481df0240e345010a47b2eca7.

yujiex committed:
add back some nullptr assertion

yujiex committed:
remove base material alloc, fix idx out of bound for new material

yujiex committed:
Move shortcut after MaterNum == 0 check

yujiex committed:
fixes: material member init after merging, new matChild chain

yujiex committed:
Merge remote-tracking branch 'origin/develop' into dynamicCoatingWallRoof

yujiex committed:
add MaterialChild Base to WindowManagerExterior*

yujiex committed:
clang-format

yujiex committed:
clang-format unit test

yujiex committed:
Fix WCE related integration tests

cannot do dynamic_cast to MaterialChild in CWCE related as the
m_MaterialProperties has a different address from the corresponding
state.dataMaterial(MaterNum) one

yujiex committed:
move thisMaterial after matid check, more assert nullptr

yujiex committed:
Merge remote-tracking branch 'origin/develop' into dynamicCoatingWallRoof

yujiex committed:
comment about why the materials need to be pre-allocated

yujiex committed:
Merge remote-tracking branch 'origin/develop' into dynamicCoatingWallRoof

yujiex committed:
Move DisplayMaterialRoughness from DataHeatBalance to Material

yujiex committed:
Move SurfaceRoughness from DataSurface to Material

yujiex committed:
Move ExtVentedCavity from DataSurfaces to DataHeatBal

yujiex committed:
Move WindowBlindProperties and Blind into Material

yujiex committed:
Move WindowComplexShade and ComplexShade

yujiex committed:
Move WindowThermalModelParams and WindowThermalModel to Material

yujiex committed:
Move SurfaceScreenProperties and SurfaceScreens to Material

yujiex committed:
Move ScreenTransData and ScreenTrans

yujiex committed:
Merge remote-tracking branch 'origin/develop' into dynamicCoatingWallRoof

amirroth commented on src/EnergyPlus/Construction.cc:
```diff
@@ -50,7 +50,6 @@
 #include <EnergyPlus/Construction.hh>
 #include <EnergyPlus/Data/EnergyPlusData.hh>
 #include <EnergyPlus/DataConversions.hh>
-#include <EnergyPlus/DataHeatBalance.hh>
```
This makes me irrationally happy.

yujiex replied:
same here :)

yujiex committed:
clang-format

yujiex committed:
Rename SurfaceScreens->Screens SurfaceScreenProperties->ScreenProperties

amirroth commented on src/EnergyPlus/DataHeatBalance.hh:
```diff
@@ -79,10 +79,8 @@ namespace DataHeatBalance {
-    using DataComplexFenestration::WindowThermalModelParams;
-    using DataSurfaces::MaxSlatAngs;
     using DataVectorTypes::Vector;
+    using Material::MaxSlatAngs;
```
Should not use `using`.

yujiex replied:
I'll remove using here

amirroth commented on src/EnergyPlus/DataSurfaces.hh:
```diff
@@ -81,7 +82,6 @@ namespace DataSurfaces {
 
     // MODULE PARAMETER DEFINITIONS:
-    constexpr int MaxSlatAngs(19);
     constexpr int MaxProfAngs(37);
```
Should this also move to `Material`? I see it used with slats.

yujiex replied:
yes, seems like they should both be in Material

amirroth commented on src/EnergyPlus/DataSurfaces.hh:
```diff
@@ -1730,7 +1654,7 @@ struct SurfacesData : BaseGlobalStruct
     EPVector<DataSurfaces::GroundSurfacesProperty> GroundSurfsProperty;
 
-    int actualMaxSlatAngs = DataSurfaces::MaxSlatAngs; // If there are no blinds in the model, then this is changed to 1 (used for shades)
+    int actualMaxSlatAngs = Material::MaxSlatAngs; // If there are no blinds in the model, then this is changed to 1 (used for shades)
```
This is the same for all blinds in the model?

yujiex replied:
yes, I believe so

amirroth commented on src/EnergyPlus/EcoRoofManager.cc:
```diff
@@ -413,15 +414,15 @@ namespace EcoRoofManager {
             }
 
-            if (RoughSurf == DataSurfaces::SurfaceRoughness::VerySmooth) {
+            if (RoughSurf == Material::SurfaceRoughness::VerySmooth) {
```
Can replace this with a `std::array` and lookup.

yujiex replied:
yeah, that's better. I'll make the Zog values into an array and look it up using roughness

amirroth commented on src/EnergyPlus/DataHeatBalance.cc:
```diff
@@ -374,8 +374,8 @@ void CheckAndSetConstructionProperties(EnergyPlusData &state,
     InsideMaterNum = thisConstruct.LayerPoint(InsideLayer);
-    auto const *thisMaterialInside = state.dataMaterial->Material(InsideMaterNum);
     if (InsideMaterNum != 0) {
+        auto const *thisMaterialInside = dynamic_cast<const Material::MaterialChild *>(state.dataMaterial->Material(InsideMaterNum));
```
I think it may be good to add `AbsorpThermal`, `AbsorpSolar`, and `AbsorpVisible` to `MaterialBase` because they are used quite often. I would leave the EMS and `VarAbsorp` stuff in `MaterialChild` though.

yujiex replied:
Maybe AbsorpThermalFront and AbsorpThermalBack should also be moved to MaterialBase with AbsorpThermal?

amirroth commented on src/EnergyPlus/Material.hh:
```diff
@@ -343,18 +384,240 @@ namespace Material {
 
     void GetMaterialData(EnergyPlusData &state, bool &errorsFound); // set to true if errors found in input
     void GetVariableAbsorptanceInput(EnergyPlusData &state, bool &errorsFound);
+    std::string DisplayMaterialRoughness(SurfaceRoughness Roughness); // Roughness String
```
This function should be replaced with `std::array<std::string_view> RoughnessNames;`

yujiex replied:
indeed

amirroth commented on src/EnergyPlus/Material.hh:
```diff
@@ -343,18 +384,240 @@ namespace Material {
+    int TotMaterials = 0;     // Total number of unique materials (layers) in this simulation
+    int TotComplexShades = 0; // Total number of shading materials for complex fenestrations
 
+    Array1D<Material::WindowBlindProperties> Blind;
```
Should this also be an EPVector?

yujiex replied:
Yeah. There's one redimension adding a new blind, but that can be done in EPVector too

amirroth commented on src/EnergyPlus/MoistureBalanceEMPDManager.cc:
```diff
@@ -121,6 +121,7 @@ Real64 CalcDepthFromPeriod(EnergyPlusData &state,
     // Calculate saturation vapor pressure at assumed temperature
     Real64 const PV_sat = Psychrometrics::PsyPsatFnTemp(state, T, "CalcDepthFromPeriod");
 
+    auto const *mat = dynamic_cast<const Material::MaterialChild *>(matBase);
```
Need `assert` here.

yujiex replied:
Added it and a few other places

amirroth commented on src/EnergyPlus/PipeHeatTransfer.cc:
```diff
@@ -658,14 +658,15 @@ void GetPipesHeatTransfer(EnergyPlusData &state)
-            state.dataPipeHT->PipeHT(Item).SoilSolarAbs = state.dataMaterial->Material(state.dataPipeHT->PipeHT(Item).SoilMaterialNum)->AbsorpSolar;
-            state.dataPipeHT->PipeHT(Item).SoilRoughness = state.dataMaterial->Material(state.dataPipeHT->PipeHT(Item).SoilMaterialNum)->Roughness;
+            auto const *thisMaterialSoil =
+                dynamic_cast<const Material::MaterialChild *>(state.dataMaterial->Material(state.dataPipeHT->PipeHT(Item).SoilMaterialNum));
```
`assert`

yujiex replied:
added here too

yujiex committed:
Change CTFCross etc from Array1D to std::array

amirroth commented on src/EnergyPlus/HeatBalanceManager.cc:
```diff
@@ -4510,73 +4510,71 @@ namespace HeatBalanceManager {
-                state.dataMaterial->Material(loop)->EMPDCoatingThickness = 0.0;
+                auto *thisMaterial = new Material::MaterialChild;
+                state.dataMaterial->Material.push_back(thisMaterial);
+                thisMaterial->Name = "";
```
If these are the default values for everything (i.e., they are what the default constructor would put in) then you don't need this code. The default constructor is called by `new`.

yujiex replied:
very true. The values are the same as the default values.

amirroth commented on src/EnergyPlus/Material.cc:
```diff
@@ -259,13 +260,14 @@ void GetMaterialData(EnergyPlusData &state, bool &ErrorsFound) // set to true if
 
             // Load the material derived type from the input data.
-            auto thisMaterial = state.dataMaterial->Material(MaterNum);
+            auto thisMaterial = new Material::MaterialChild;
```
I know that plain `auto` works here, but please use `auto *` because otherwise it looks like something that should be an `auto &` if you know what I mean.

yujiex replied:
yeah indeed. Forgot here. I'll change to auto *

amirroth commented on src/EnergyPlus/Material.cc:
```diff
@@ -291,14 +293,15 @@ void GetMaterialData(EnergyPlusData &state, bool &ErrorsFound) // set to true if
         thisMaterial->Density = 2240.0;    // kg/m3
         thisMaterial->SpecHeat = 900.0;    // J/kgK
-        thisMaterial->Roughness = DataSurfaces::SurfaceRoughness::MediumRough;
+        thisMaterial->Roughness = Material::SurfaceRoughness::MediumRough;
```
Don't need `Material::` namespace scoping when you are inside that namespace.

yujiex replied:
right right

amirroth commented on src/EnergyPlus/Material.cc:
```diff
@@ -454,7 +460,8 @@ void GetMaterialData(EnergyPlusData &state, bool &ErrorsFound) // set to true if
-        auto *thisMaterial = state.dataMaterial->Material(MaterNum);
+        auto *thisMaterial = new Material::MaterialChild;
+        state.dataMaterial->Material(MaterNum) = thisMaterial;
         thisMaterial->Group = MaterialGroup::IRTMaterial;
```
Probably don't need to have the word `Material` in type names and values, e.g., `MaterialGroup::IRTMaterial` should `Group::IRT` (and `Material::Group::IRT` outside of the `Material` namespace). I would also expand IRT to something more meaningful like `IRTransparent`.

yujiex replied:
true. will remove these `Material` words

amirroth commented on src/EnergyPlus/Material.cc:
```diff
@@ -1710,7 +1727,7 @@ void GetMaterialData(EnergyPlusData &state, bool &ErrorsFound) // set to true if
                                   "\", must be one of DoNotModel, ModelAsDirectBeam or ModelAsDiffuse.");
         }
-        thisMaterial->Roughness = DataSurfaces::SurfaceRoughness::MediumRough;
+        thisMaterial->Roughness = Material::SurfaceRoughness::MediumRough;
```
Again, in general don't need `namespaceX::` when you are inside that namespace.

yujiex replied:
removed it here as well

yujiex committed:
remove using Material::MaxSlatAngs

yujiex committed:
Rename SurfaceRoughness to Roughness

yujiex committed:
Move MaxProfAngs to Material

yujiex committed:
look up Zog in an array using roughness

yujiex committed:
Replace DisplayMaterialRoughness with RoughnessNames lookup array

yujiex committed:
Change Blind from Array1D to EPVector

yujiex committed:
more nullptr assertions after dynamic_cast

yujiex committed:
auto thisMaterial changed to auto *thisMaterial

yujiex committed:
Remove value overwrite as they are the same as default constructor

yujiex committed:
Remove Material::  in Material.cc

yujiex committed:
Rename "Material" in MaterialGroup, IRTMaterial, RegularMaterial

yujiex committed:
Rename IRT to IRTransparent

yujiex committed:
Merge remote-tracking branch 'origin/develop' into dynamicCoatingWallRoof

yujiex committed:
Revert "Rename SurfaceRoughness to Roughness"

This reverts commit 47faad50645387fb6b2419f6a59b53d29e4fa0ce.

yujiex committed:
fix license text and clang-format

yujiex committed:
fix messed-up indexing from 5dd135397a

amirroth commented on src/EnergyPlus/Construction.hh:
```diff
@@ -52,9 +52,9 @@
 #include <EnergyPlus/Data/BaseData.hh>
 #include <EnergyPlus/DataBSDFWindow.hh>
 #include <EnergyPlus/DataHeatBalance.hh>
-#include <EnergyPlus/DataSurfaces.hh>
```
This is also very good. Generally speaking modules should include headers of modules that they logically contain. A construction logically contains materials so it is proper for the construction module to include `material.hh`. But constructions do not logically contain surfaces, it's the other way around. So the construction module should not include `surface.hh`, if anything it should be the other way around. If you find these types of include "inversions" it generally means that someone put something in the wrong place.

yujiex replied:
That's a nice way of reasoning about which item goes into which file

amirroth commented on src/EnergyPlus/EcoRoofManager.cc:
```diff
@@ -413,19 +414,9 @@ namespace EcoRoofManager {
-                state.dataEcoRoofMgr->Zog = 0.005;
-            } // TODO: fix this after creating FindEnumeratedValueIndex()
+            // "VeryRough, Rough, MediumRough, MediumSmooth, Smooth, VerySmooth"
+            std::array<Real64, 6> zogLookup = {0.005, 0.0030, 0.0020, 0.0015, 0.0010, 0.0008};
```
Small comment. Replace `6` with `static_cast<int>(Material::SurfaceRoughness::Num)`.

yujiex replied:
Right right, for sure

yujiex committed:
Add unit test for eoshiftArray and fix its bug

yujiex committed:
change array length hard code to SurfaceRoughness::Num

yujiex committed:
clang-format

yujiex committed:
Merge remote-tracking branch 'origin/develop' into dynamicCoatingWallRoof

yujiex committed:
Merge remote-tracking branch 'origin/develop' into dynamicCoatingWallRoof

yujiex committed:
fix regression diff, use fill to assign to std::array

array = {init_val} only affects the first item. need to use array.fill()
to assign to all elements

yujiex committed:
clang-format

yujiex committed:
fix varname: somehow strangely turned into lower case

yujiex committed:
Merge remote-tracking branch 'origin/develop' into dynamicCoatingWallRoof

yujiex committed:
Merge remote-tracking branch 'origin/develop' into dynamicCoatingWallRoof

yujiex committed:
Merge remote-tracking branch 'origin/develop' into dynamicCoatingWallRoof

yujiex committed:
add value initialization to std arrays

yujiex committed:
Merge remote-tracking branch 'origin/develop' into dynamicCoatingWallRoof

yujiex committed:
clang-format

Myoldmopar commented on src/EnergyPlus/Construction.cc:
```diff
@@ -214,7 +213,8 @@ void ConstructionProps::calculateTransferFunction(EnergyPlusData &state, bool &E
 
-        auto *thisMaterial = state.dataMaterial->Material(CurrentLayer);
+        auto *thisMaterial = dynamic_cast<Material::MaterialChild *>(state.dataMaterial->Material(CurrentLayer));
+        assert(thisMaterial != nullptr);
```
Asserting _after_ the dynamic cast looks a little funny, but it's fine.  If the original material layer is null, then this dynamic cast will just point to the same null address without complaint, so it's fine.

My bigger question is about the dynamic_cast at all.  There is certainly a place for it, but I would encourage you to think twice or thrice about using it.  These things are often a sign of some member variables/methods that need to be inherited/overridden.  Anywho, I'll just drop that comment here for now.

yujiex replied:
For nullptr checks before the dynamic_cast, should it be at the `new MaterialBase`?

the nullptr check after the dynamic_cast her is to catch state.dataMaterial->Material(CurrentLayer) being a different type then MaterialChild

Myoldmopar replied:
I don't think you need to worry about the dynamic cast for an incompatible type.  https://godbolt.org/z/qoKj1s8MM

As I understand it, the compiler will complain if the target type and the source type are not polymorphic with each other.  Perhaps I am missing a corner case...  I assumed the check was in case the Material(CurrentLayer) was null.

amirroth replied:
@Myoldmopar, you're half right here. Let's say you have four classes: `Base`, `Base2`, `Child1 : Base`, `Child2 : Base`. If you have a pointer of type `Base *`, the compiler will throw an error if you try to `dynamic_cast<Base2 *>` and not if you `dynamic_cast<Child1 *>` or `dynamic_cast<Child 2 *>`. But the compiler can't do anything if you `dynamic_cast<Child1 *>` and the object actually turns out to be of type `Child2`. In that case the C++ runtime returns `nullptr`. That's the case you have to `assert` for.

Myoldmopar commented on src/EnergyPlus/Construction.cc:
```diff
@@ -458,10 +458,10 @@ void ConstructionProps::calculateTransferFunction(EnergyPlusData &state, bool &E
-                        this->CTFCross(HistTerm) = otherConstruction.CTFCross(HistTerm);
-                        this->CTFOutside(HistTerm) = otherConstruction.CTFInside(HistTerm);
-                        if (HistTerm != 0) this->CTFFlux(HistTerm) = otherConstruction.CTFFlux(HistTerm);
+                        this->CTFInside[HistTerm] = otherConstruction.CTFOutside[HistTerm];
```
Just so happy to see proper array access here.

Myoldmopar commented on src/EnergyPlus/Construction.cc:
```diff
@@ -1898,19 +1898,19 @@ void ConstructionProps::reportTransferFunction(EnergyPlusData &state, int const
-        switch (thisMaterial->Group) {
-        case Material::MaterialGroup::Air: {
+        switch (thisMaterial->group) {
+        case Material::Group::Air: {
```
Much cleaner, 👍

Myoldmopar commented on src/EnergyPlus/Construction.hh:
```diff
@@ -88,36 +88,36 @@ namespace Construction {
-        Array1D<Real64> CTFSourceIn;  // Heat source/sink inside terms of CTF equation
-        Array1D<Real64> CTFSourceOut; // Heat source/sink outside terms of CTF equation
-        Real64 CTFTimeStep = 0.0;     // Time increment for stable simulation of construct (could be greater than TimeStep)
+        std::array<Real64, MaxCTFTerms> CTFCross{};     // Cross or Y terms of the CTF equation
```
👍 👍

Myoldmopar commented on src/EnergyPlus/Construction.hh:
```diff
@@ -271,17 +271,12 @@ namespace Construction {
-              BlTransDiffVis(DataSurfaces::MaxSlatAngs, 0.0), BlReflectSolDiffBack(DataSurfaces::MaxSlatAngs, 0.0),
-              BlReflectSolDiffFront(DataSurfaces::MaxSlatAngs, 0.0), BlReflectVisDiffBack(DataSurfaces::MaxSlatAngs, 0.0),
-              BlReflectVisDiffFront(DataSurfaces::MaxSlatAngs, 0.0), TransSolBeamCoef(6, 0.0), TransVisBeamCoef(6, 0.0), ReflSolBeamFrontCoef(6, 0.0),
+            : LayerPoint(MaxLayersInConstruct, 0), AbsDiffBlind(Material::MaxSlatAngs, 0.0), AbsDiffBlindGnd(Material::MaxSlatAngs, 0.0),
```
It will be great to continue cleaning this up as more members are converted to arrays.

yujiex replied:
Yeah, for sure. I did the ones that are 0-indexing first.

Myoldmopar commented on src/EnergyPlus/DataHeatBalance.hh:
```diff
@@ -905,6 +902,55 @@ namespace DataHeatBalance {
+        Real64 PassiveMdotTherm; // Nat. Vent air change rate from buoyancy-driven flow [kg/s]
+
+        // Default Constructor
+        ExtVentedCavityStruct()
```
Missed opportunity to move the member initialization to the member declarations, but not a big deal.

yujiex replied:
Oh, right, I can change the initialization to member declaration.

Myoldmopar replied:
You could, but I wouldn't worry about it if there's nothing else to change here.

Myoldmopar commented on src/EnergyPlus/DataHeatBalance.cc:
```diff
@@ -400,29 +400,27 @@ void CheckAndSetConstructionProperties(EnergyPlusData &state,
     }
 
     if (InsideMaterNum == 0) return;
+    auto const *thisMaterialInside = dynamic_cast<const Material::MaterialChild *>(state.dataMaterial->Material(InsideMaterNum));
```
All this dynamic casting makes me think that a next iteration of this work should do something like this:

```c++
struct MaterialBase {
  std::string name;
  virtual void applyProperty() {} // assuming the base class doesn't need to do anything
};

struct MaterialDerived : MaterialBase {
  Real64 someDerivedProperty;
  virtual void applyProperty() override;
};
```

then in the cc file:

```c++
void MaterialDerived::applyProperty() {
  this->something = someDerivedProperty ** 2;
}
```

This way you can just call `applyProperty` on the instance, and it will automatically use the right version regardless of which type of actual material you have.

yujiex replied:
Does this mean when we want to get a member of a struct, we'll need to have a get and a set virtual function for every member of every derived struct?

Myoldmopar replied:
It means we will want to think about what we actually need the material classes to do/provide.  Class inheritance structures should provide layers of abstraction.  I'm not proposing we want more than one layer of inheritance here, but even in this one layer, we should still find ways to abstract the functionality and data.  Let's say you need to calculate F for a material.  In one material, F is a function of 2 material properties: x and y.   In another material, F is a function of 3 material properties: x, y, and z.  Out in client code, you don't want to constantly dynamic cast instances to each derived type and grab individual material properties and using them to calculate F.  No, you want to create a pure virtual function to calculate F in the base class.   Then each derived material type must implement their own function to calculate F.  And if x and y are actually shared by all the material types, those should be stored in the base class as well, and each derived type only needs to add data specific to their type.  This tightens up the interface between the material class and the clients using it and encourages code reuse, data hiding, and proper ownership.

This PR is a great step forward for sure.  Most of what I am saying here is just me being optimistic about where this will go next.  Thanks @yujiex

Myoldmopar commented on src/EnergyPlus/DataHeatBalance.cc:
```diff
@@ -704,52 +695,51 @@ void CheckAndSetConstructionProperties(EnergyPlusData &state,
     } else { // Opaque surface
-        thisConstruct.InsideAbsorpThermal = state.dataMaterial->Material(thisConstruct.LayerPoint(InsideLayer))->AbsorpThermal;
-        thisConstruct.OutsideAbsorpThermal = state.dataMaterial->Material(thisConstruct.LayerPoint(1))->AbsorpThermal;
+        thisConstruct.InsideAbsorpThermal = thisMaterialInside->AbsorpThermal;
```
The object shortcut is certainly nice.

Myoldmopar commented on src/EnergyPlus/Material.hh:
```diff
@@ -145,31 +161,58 @@ namespace Material {
+        }
+    };
+
+    struct MaterialChild : public MaterialBase
```
MaterialChild feels like the wrong name.  If we are going to use class inheritance, the names should be more meaningful.  

```c++
struct MaterialBase {};
struct MaterialRegular : MaterialBase {};
struct MaterialSomething : MaterialBase {};
```

yujiex replied:
This is the second step of splitting. After this, we'll subdivide MaterialChild into more meaningful classes like gas/gas mixture, simple material, complex opaque, complex window, potentially some other types

Myoldmopar replied:
Yep, that's a great path forward.

Myoldmopar commented on src/EnergyPlus/Material.hh:
```diff
@@ -343,6 +388,221 @@ namespace Material {
+        virtual bool dummy()
+        {
+            return true;
+        }
```
dummy?

yujiex replied:
I need a virtual function to do dynamic_cast

Myoldmopar replied:
Got it now.  Thanks.  I can't believe the compiler will actually stop if it doesn't find a virtual function.  I assumed it would be fine.  Of course this will go away as things in the base class become virtual and common capabilities are pulled into the base class.

Myoldmopar reviewed (commented):
This is obviously a great step forward.  I have a few thoughts about the new structure with respect to naming, member methods, and the uses of dynamic casts.  In general, I think those can all be accepted as is, and a next (final?) cleanup pass could really polish this up nicely.

There is some virtual dummy() function in there, which ... as long as it's not being used I guess it's not hurting anything.  

eoshiftArray is only ever called within EnergyPlus with a shift of -1.  I'm wondering if it's worthwhile to eliminate that argument entirely in that function.  It could clean it up and remove an argument.

Let me know what you think about those two last pieces.  Either way this is very close.

yujiex commented:
> This is obviously a great step forward. I have a few thoughts about the new structure with respect to naming, member methods, and the uses of dynamic casts. In general, I think those can all be accepted as is, and a next (final?) cleanup pass could really polish this up nicely.
> 
> There is some virtual dummy() function in there, which ... as long as it's not being used I guess it's not hurting anything.
> 
> eoshiftArray is only ever called within EnergyPlus with a shift of -1. I'm wondering if it's worthwhile to eliminate that argument entirely in that function. It could clean it up and remove an argument.
> 
> Let me know what you think about those two last pieces. Either way this is very close.

Hi @Myoldmopar, re these comments

Definitely, there are at least one or two more steps to further subdivide the MaterialChild and polish it more. 

The virtual dummy() function is because the compiler will complain if I don't have a virtual function when I do dynamic_cast.  So it's pretty much just to make the compiler happy. No actual use for now.

eoshiftArray can definitely eliminate the shift argument. Although would you foresee other future migrations from Array1D to std::array that will need such a shift function? I'm undecided whether it should be very specific or very general

amirroth commented:
> Definitely, there are at least one or two more steps to further subdivide the MaterialChild and polish it more.

@Myoldmopar, Yujie also took the opportunity to move some Material content from various other modules (mostly DataHeatBalance) into the Material module. This helps simplify and reduce the "include graph" and will also help in constructing the final Material object hierarchy. Many of these stragglers (e.g., Blind, Shade, Complex Shade) are actually types of materials, but were not added to the central Material object because someone wisely noted that that thing was already getting too big. Instead, these these are separate objects that point to Material and vice versa (gross). The plan is to make them into sub-classes and eliminate the need for this type of cross-referencing.

> The virtual dummy() function is because the compiler will complain if I don't have a virtual function when I do dynamic_cast. So it's pretty much just to make the compiler happy. No actual use for now.

That's funny. I didn't know that this happens!

Myoldmopar commented:
> That's funny. I didn't know that this happens!

It doesn't just complain about it, it actually aborts!?  Uncomment the virtual function to see it compile happily. 
 https://godbolt.org/z/EKffsY7qd . I had no idea.

yujiex commented:
> > That's funny. I didn't know that this happens!
> 
> It doesn't just complain about it, it actually aborts!? Uncomment the virtual function to see it compile happily. https://godbolt.org/z/EKffsY7qd . I had no idea.

a lot of surprises for sure

Myoldmopar commented:
OK, no need to dwell on this any further, this is good.  Thanks @yujiex for the effort, and @amirroth for the guidance and reviewing.  Merging this one in, any straggling pieces can be done in a following effort after release.

yujiex commented:
> OK, no need to dwell on this any further, this is good. Thanks @yujiex for the effort, and @amirroth for the guidance and reviewing. Merging this one in, any straggling pieces can be done in a following effort after release.

Thanks, @Myoldmopar and @amirroth for the reviewing and guidance :) I learned a lot of new things through the process
